File: ios/LuminaLog/Core/Crypto/WrappedKey.swift
import Foundation
import CryptoKit

enum WrappedKeyError: LocalizedError {
    case unwrapFailed

    var errorDescription: String? {
        switch self {
        case .unwrapFailed: return "Could not unwrap the encryption key."
        }
    }
}

/// A Data Encryption Key (DEK) wrapped (encrypted) under some Key Encryption Key
/// (KEK). Mirrors the server `WrappedDEK` shape `{ v, iv, ct, tag }` with base64
/// blobs (see `server/src/crypto/keyService.ts`), so a wrap produced on-device
/// round-trips through Firestore and stays interoperable with the server model.
///
/// Pure value type. AES-256-GCM is used with NO additional authenticated data,
/// matching the server's `wrapDEK`/`unwrapDEK` so a wrap is portable either way.
struct WrappedKey: Equatable {

    static let version = 1

    let v: Int
    let iv: Data   // 12-byte GCM nonce
    let ct: Data   // wrapped-DEK ciphertext
    let tag: Data  // 16-byte GCM tag

    var firestoreData: [String: Any] {
        [
            "v": v,
            "iv": iv.base64EncodedString(),
            "ct": ct.base64EncodedString(),
            "tag": tag.base64EncodedString(),
        ]
    }

    init(iv: Data, ct: Data, tag: Data, v: Int = WrappedKey.version) {
        self.v = v
        self.iv = iv
        self.ct = ct
        self.tag = tag
    }

    /// Parse from a Firestore value. Returns nil for anything that is not a
    /// well-formed v1 wrap.
    init?(data: Any?) {
        guard
            let dict = data as? [String: Any],
            let v = dict["v"] as? Int, v == Self.version,
            let ivB64 = dict["iv"] as? String, let iv = Data(base64Encoded: ivB64),
            let ctB64 = dict["ct"] as? String, let ct = Data(base64Encoded: ctB64),
            let tagB64 = dict["tag"] as? String, let tag = Data(base64Encoded: tagB64),
            // Reject malformed envelopes early: 12-byte GCM nonce, 16-byte tag.
            iv.count == 12, tag.count == 16, !ct.isEmpty
        else { return nil }
        self.v = v
        self.iv = iv
        self.ct = ct
        self.tag = tag
    }
}

extension WrappedKey {

    /// Wrap a 256-bit DEK under a KEK using AES-256-GCM (no AAD).
    /// Seal with a fresh nonce + a valid 256-bit key never fails, so this is
    /// non-throwing; a failure here would be a programmer error, not runtime.
    static func wrapping(dek: SymmetricKey, under kek: SymmetricKey) -> WrappedKey {
        let nonce = AES.GCM.Nonce()
        // Force-unwrap is safe: fresh nonce + valid key + small plaintext.
        let sealed = try! AES.GCM.seal(dek.rawData, using: kek, nonce: nonce)
        return WrappedKey(iv: Data(nonce), ct: sealed.ciphertext, tag: sealed.tag)
    }

    /// Unwrap this wrap back into a DEK under `kek`. Fails closed: any tag
    /// mismatch (wrong KEK / tampered blob) or wrong key length throws.
    func unwrapping(under kek: SymmetricKey) throws -> SymmetricKey {
        do {
            let box = try AES.GCM.SealedBox(
                nonce: AES.GCM.Nonce(data: iv),
                ciphertext: ct,
                tag: tag
            )
            let raw = try AES.GCM.open(box, using: kek)
            guard raw.count == 32 else { throw WrappedKeyError.unwrapFailed }
            return SymmetricKey(data: raw)
        } catch {
            throw WrappedKeyError.unwrapFailed
        }
    }
}

extension SymmetricKey {
    /// The raw key bytes as `Data`.
    var rawData: Data { withUnsafeBytes { Data($0) } }
}

File: web/src/lib/crypto/wrappedKey.ts
// The `WrappedKey` envelope — `{v,iv,ct,tag}`, AES-256-GCM, **no AAD** — byte-
// compatible with iOS `WrappedKey.swift` (`firestoreData`) and the server
// `WrappedDEK`. This is DISTINCT from the field-encryption envelope in
// `envelope.ts` (`{v,alg,iv,ct,tag}`, AAD-bound): the wrapped-key/vector path
// uses no AAD and no `alg` field, matching CryptoKit's `AES.GCM.seal(..)` with a
// fresh nonce. Reused for both the DEK wrap (passkey-PRF) and the vector wrap.

const VERSION = 1
const IV_BYTES = 12
const TAG_BYTES = 16

export interface WrappedKeyEnvelope {
  v: 1
  iv: string // base64, 12-byte GCM nonce
  ct: string // base64 ciphertext
  tag: string // base64, 16-byte GCM tag
}

function toBase64(bytes: Uint8Array): string {
  let binary = ''
  for (let i = 0; i < bytes.length; i++) binary += String.fromCharCode(bytes[i])
  return btoa(binary)
}

function fromBase64(b64: string): Uint8Array {
  const binary = atob(b64)
  const out = new Uint8Array(binary.length)
  for (let i = 0; i < binary.length; i++) out[i] = binary.charCodeAt(i)
  return out
}

export function isWrappedKeyEnvelope(value: unknown): value is WrappedKeyEnvelope {
  if (typeof value !== 'object' || value === null) return false
  const e = value as Record<string, unknown>
  if (e.v !== VERSION) return false
  if (typeof e.iv !== 'string' || typeof e.ct !== 'string' || typeof e.tag !== 'string') return false
  try {
    return fromBase64(e.iv).length === IV_BYTES && fromBase64(e.tag).length === TAG_BYTES && fromBase64(e.ct).length > 0
  } catch {
    return false
  }
}

/** Seal `plaintext` under `key` (AES-256-GCM, fresh 12-byte nonce, no AAD). */
export async function wrap(key: CryptoKey, plaintext: Uint8Array): Promise<WrappedKeyEnvelope> {
  const iv = crypto.getRandomValues(new Uint8Array(IV_BYTES))
  const combined = new Uint8Array(await crypto.subtle.encrypt({ name: 'AES-GCM', iv }, key, plaintext))
  // WebCrypto returns ciphertext||tag; split off the trailing 16-byte tag to
  // match the CryptoKit `{ct, tag}` split.
  const ct = combined.slice(0, combined.length - TAG_BYTES)
  const tag = combined.slice(combined.length - TAG_BYTES)
  return { v: VERSION, iv: toBase64(iv), ct: toBase64(ct), tag: toBase64(tag) }
}

/** Open a `WrappedKey` envelope under `key`. Fails closed on tamper / wrong key / malformed. */
export async function unwrap(key: CryptoKey, env: WrappedKeyEnvelope): Promise<Uint8Array> {
  if (!isWrappedKeyEnvelope(env)) throw new Error('Malformed WrappedKey envelope')
  const iv = fromBase64(env.iv)
  const ct = fromBase64(env.ct)
  const tag = fromBase64(env.tag)
  const combined = new Uint8Array(ct.length + tag.length)
  combined.set(ct, 0)
  combined.set(tag, ct.length)
  const plaintext = await crypto.subtle.decrypt({ name: 'AES-GCM', iv }, key, combined)
  return new Uint8Array(plaintext)
}
